platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/exec/CleanUpVirtualFileSystemAfterBuild.java

```diff
@@ -0,0 +1,102 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.launcher.daemon.server.exec;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.internal.buildoption.DefaultInternalOptions;
+import org.gradle.internal.buildoption.InternalOptions;
+import org.gradle.internal.concurrent.ExecutorFactory;
+import org.gradle.internal.concurrent.ManagedExecutor;
+import org.gradle.internal.concurrent.Stoppable;
+import org.gradle.internal.service.scopes.GradleUserHomeScopeServiceRegistry;
+import org.gradle.internal.service.scopes.VirtualFileSystemServices;
+import org.gradle.internal.vfs.VirtualFileSystem;
+import org.gradle.internal.watch.vfs.BuildLifecycleAwareVirtualFileSystem;
+import org.gradle.launcher.daemon.protocol.Build;
+import org.gradle.launcher.daemon.server.api.DaemonCommandExecution;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Asynchronously cleans up the VFS after a build.
+ *
+ * Unblocks the client to receive the build finished event while the cleanup is happening.
+ * However, the next build is not allowed to start until the cleanup is finished.
+ */
+@NonNullApi
+public class CleanUpVirtualFileSystemAfterBuild extends BuildCommandOnly implements Stoppable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CleanUpVirtualFileSystemAfterBuild.class);
+
+    private final GradleUserHomeScopeServiceRegistry userHomeServiceRegistry;
+    private final ManagedExecutor executor;
+
+    private CompletableFuture<Void> pendingCleanup = CompletableFuture.completedFuture(null);
+
+    public CleanUpVirtualFileSystemAfterBuild(ExecutorFactory executorFactory, GradleUserHomeScopeServiceRegistry userHomeServiceRegistry) {
+        this.executor = executorFactory.create("VFS cleanup");
+        this.userHomeServiceRegistry = userHomeServiceRegistry;
+    }
+
+    @Override
+    protected void doBuild(DaemonCommandExecution execution, Build build) {
+        waitForPendingCleanupToFinish();
+        try {
+            execution.proceed();
+        } finally {
+            startAsyncCleanupAfterBuild(build);
+        }
+    }
+
+    private void startAsyncCleanupAfterBuild(Build build) {
+        userHomeServiceRegistry.getCurrentServices().ifPresent(serviceRegistry ->
+            pendingCleanup = CompletableFuture.runAsync(() -> {
```

Comment:
🤔 This looks somewhat hacky. Shouldn't we have already some infrastructure that allows us running stuff between builds? I am in particular worried here that we use the common pool instead of one of Gradle's executor pools to run the cleanup action. Moreover, the mutable state to hand the future from the last build to the current build seems strange.
Do we do anything like this somewhere else in Gradle?

Reply:
> 🤔 This looks somewhat hacky. Shouldn't we have already some infrastructure that allows us running stuff between builds?

I did some research and asked people about an "in-between builds" infrastructure, but nobody had an idea for it.


> I am in particular worried here that we use the common pool instead of one of Gradle's executor pools to run the cleanup action.

We are not using the common pool. Instead, we are creating a single-threaded managed pool, see the constructor.

> Moreover, the mutable state to hand the future from the last build to the current build seems strange.

I see what you mean. I tried to make it more obvious how the future is supposed to work, PTAL.

Note that this class is not thread safe, that's why it should be okay to handle the future in this way.

> Do we do anything like this somewhere else in Gradle?

I'm not aware of another example of doing something like this in Gradle.